File: app/src/main/java/com/app/missednotificationsreminder/service/RemindJob.java
package com.app.missednotificationsreminder.service;

import com.app.missednotificationsreminder.di.Injector;
import com.app.missednotificationsreminder.service.event.RemindEvents;
import com.app.missednotificationsreminder.util.event.RxEventBus;
import com.evernote.android.job.Job;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import dagger.android.AndroidInjector;
import dagger.android.ContributesAndroidInjector;
import rx.Completable;
import timber.log.Timber;

/**
 * The remind job
 */

public class RemindJob extends Job {

    public static final String TAG = "REMIND_JOB";

    @Inject RxEventBus mEventBus;

    @NonNull @Override protected Result onRunJob(@NonNull Params params) {
        Timber.d("onRunJob() called with: params = %s",
                params);
        // inject dependencies
        AndroidInjector appGraph = Injector.obtain(getContext().getApplicationContext());
        appGraph.inject(this);

        // request reminder and await for the reminder completed event
        mEventBus.toObserverable()
                .filter(event -> event == RemindEvents.REMINDER_COMPLETED)
                .mergeWith(Completable.fromAction(() -> mEventBus.send(RemindEvents.REMIND)).toObservable())
                .toBlocking()
                .first();
        Timber.d("onRunJob() done");
        return Result.SUCCESS;
    }

    @dagger.Module
    public static abstract class Module {
        @ContributesAndroidInjector
        abstract RemindJob contribute();
    }
}

File: app/src/main/java/com/app/missednotificationsreminder/service/RemindJob.kt
package com.app.missednotificationsreminder.service

import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.app.missednotificationsreminder.di.Injector.Companion.obtain
import com.app.missednotificationsreminder.service.event.RemindEvents
import com.app.missednotificationsreminder.util.event.FlowEventBus
import dagger.android.AndroidInjector
import dagger.android.ContributesAndroidInjector
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.*
import timber.log.Timber
import javax.inject.Inject

/**
 * The remind job
 */
@OptIn(FlowPreview::class, ExperimentalCoroutinesApi::class)
class RemindJob(context: Context, params: WorkerParameters)
    : CoroutineWorker(context, params) {
    init {
        Timber.d("RemindJob() called with: params = %s",
                params)
        // inject dependencies
        val appGraph: AndroidInjector<Any>? = obtain(context.applicationContext)
        appGraph!!.inject(this)
    }

    @Inject
    lateinit var mEventBus: FlowEventBus

    override suspend fun doWork(): Result = coroutineScope {
        Timber.d("doWork() called")
        mEventBus.toFlow()
                .filter { event -> event === RemindEvents.REMINDER_COMPLETED }
                .onStart {  mEventBus.send(RemindEvents.REMIND)  }
                .first()
        Timber.d("doWork() done")
        Result.success()
    }

    @dagger.Module
    abstract class Module {
        @ContributesAndroidInjector
        abstract fun contribute(): RemindJob
    }

    companion object {
        const val TAG = "REMIND_JOB"
    }
}